src/state/actions/annotation.js

```diff
@@ -50,6 +52,105 @@ export function receiveAnnotationFailure(targetId, annotationId, error) {
+ * TODO: also handle v3 target attribute
+ */
+function coerceAnnotationToCanvasId(targetId, annotationJson) {
+  const coercedJson = Object.assign({}, annotationJson);
+
+  coercedJson.resources = annotationJson.resources.map((resource) => {
+    const coercedResource = Object.assign({}, resource);
+    coercedResource.on = [].concat(coercedResource.on);
+    if (coercedResource.on[0].includes('xywh=')) {
+      coercedResource.on[0] = coercedResource.on[0].replace(/^[^#]+/, targetId); // replace up to hash
+    }
+    return coercedResource;
+  });
+
+  return coercedJson;
+}
+
+/**
+ * filterAnnotationResources - filter annotation resources
+ *
+ * @param  {Object} annotationJson
+ * TODO: make this into a configuration option
+ */
+function filterAnnotationResources(annotationJson) {
+  const filteredJson = Object.assign({}, annotationJson);
+
+  filteredJson.resources = annotationJson.resources.filter(resource => resource.dcType === 'Line');
+
+  return filteredJson;
+}
+
+/**
+ * fetchAnnotationResourcesFulltext - fetch annotation fulltext
+ *
+ * @param  {Object} annotationJson
+ */
+function fetchAnnotationResourcesFulltext(annotationJson) {
+  const urls = annotationJson.resources
+    .filter(resource => !resource.resource.chars && resource.resource['@id'])
+    .map(resource => resource.resource['@id'].split('#')[0]);
+
+  const fulltext = {};
+
+  // TODO: error handling
+  const fetches = uniq(urls).map(url => axios.get(url)
+    .then((response) => {
+      if (response.data.type === 'FullTextResource') fulltext[url] = response.data.value;
+    }));
+
+  return Promise.all(fetches).then(() => fulltext);
+}
+
+/**
+ * filterAnnotationResources - dereference annotation resources
+ *
+ * @param  {Object} annotationJson
+ */
+function dereferenceAnnotationResources(annotationJson) {
+  return fetchAnnotationResourcesFulltext(annotationJson)
+    .then((fulltext) => {
+      // console.log('fulltext', fulltext);
+      const dereferencedJson = Object.assign({}, annotationJson);
+
+      dereferencedJson.resources = annotationJson.resources.map((resource) => {
+        const dereferencedResource = Object.assign({}, resource);
+        if (dereferencedResource.resource.chars || !dereferencedResource.resource['@id']) {
+          return dereferencedResource;
+        }
+
+        const url = dereferencedResource.resource['@id'].split('#')[0];
+        if (!fulltext[url]) return dereferencedResource;
+
+        const fragment = dereferencedResource.resource['@id'].split('#')[1];
+
+        dereferencedResource.resource.chars = fulltext[url];
+
+        if (fragment) {
+          const charMatch = fragment.match(/char=(\d+),(\d+)$/);
+          if (charMatch) {
+            dereferencedResource.resource.chars = dereferencedResource.resource.chars.slice(
+              Number(charMatch[1]),
+              Number(charMatch[2]) + 1,
+            );
+            // console.log('annotation chars', dereferencedResource.resource.chars);
```

Comment:
remove console.log

Reply:
Deliberately left in for easy restoration to debug while WIP. :)

Reply:
ok fair enough :)